SU2_CFD/include/solvers/CFVMFlowSolverBase.hpp

```diff
@@ -315,6 +316,39 @@ class CFVMFlowSolverBase : public CSolver {
     }
   }
 
+
+inline void Custom_Source_Residual(CGeometry *geometry, CSolver **solver_container,
+                                   CNumerics **numerics_container, CConfig *config, unsigned short iMesh) {
+
+  /*--- Pick one numerics object per thread. ---*/
+  CNumerics* numerics = numerics_container[SOURCE_SECOND_TERM + omp_get_thread_num()*MAX_TERMS];
+
+  unsigned short iVar;
+  unsigned long iPoint;
+  AD::StartNoSharedReading();
+
+  SU2_OMP_FOR_STAT(omp_chunk_size)
+  for (iPoint = 0; iPoint < nPointDomain; iPoint++) {
+
+    /*--- Load the volume of the dual mesh cell ---*/
+
+    numerics->SetVolume(geometry->nodes->GetVolume(iPoint));
+
+    /*--- Get control volume size. ---*/
+    su2double Volume = geometry->nodes->GetVolume(iPoint);
+
+    /*--- Compute the residual for this control volume and subtract. ---*/
+    for (iVar = 0; iVar < nVar; iVar++) {
+      LinSysRes[iPoint*nVar+iVar] += PointSource[iPoint][iVar] * Volume;
```

Comment:
We should avoid assumptions regarding how the data is stored internally and let the classes deal with that.
```suggestion
      LinSysRes(iPoint, iVar) += PointSource(iPoint, iVar) * Volume;
```